ci-bench/src/bench_lib/cachegrind.rs

```diff
@@ -0,0 +1,237 @@
+    fn ensure_cachegrind_available() -> anyhow::Result<()> {
+        let result = Command::new("valgrind")
+            .arg("--tool=cachegrind")
+            .arg("--version")
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .status();
+
+        match result {
+            Err(e) => anyhow::bail!("Unexpected error while launching cachegrind. Error: {}", e),
+            Ok(status) => {
+                if status.success() {
+                    Ok(())
+                } else {
+                    anyhow::bail!("Failed to launch cachegrind. Error: {}. Please ensure that valgrind is installed and on the $PATH.", status)
+                }
+            }
+        }
+    }
+
+    /// See docs for [`Self::run_bench`]
+    fn run_bench_side(
+        arch: &str,
+        executable: &str,
+        benchmark_index: u32,
+        side: Side,
+        name: &str,
+        stdin: Stdio,
+        stdout: Stdio,
+    ) -> anyhow::Result<BenchSubprocess> {
+        let output_file = PathBuf::from(format!("target/cachegrind/cachegrind.out.{}", name));
+        std::fs::create_dir_all(output_file.parent().unwrap())
+            .context("Failed to create cachegrind output directory")?;
+
+        // Run under setarch to disable ASLR, to reduce noise
+        let mut cmd = Command::new("setarch");
+        let child = cmd
+            .arg(arch)
+            .arg("-R")
+            .arg("valgrind")
+            .arg("--tool=cachegrind")
+            .arg("--cache-sim=no")
+            .arg("--log-file=/dev/null")
+            .arg(format!("--cachegrind-out-file={}", output_file.display()))
+            .arg(executable)
+            .arg("run-single")
+            .arg(benchmark_index.to_string())
+            .arg(side.as_str())
+            .stdin(stdin)
+            .stdout(stdout)
+            .stderr(Stdio::inherit())
+            .spawn()
+            .context("Failed to run benchmark in cachegrind")?;
+
+        Ok(BenchSubprocess {
+            process: child,
+            output_file,
+        })
+    }
+}
+
+/// A running subprocess for one of the sides of the benchmark (client or server)
+struct BenchSubprocess {
+    /// The benchmark's child process, running under cachegrind
+    process: Child,
+    /// Cachegrind's output file for this benchmark
+    output_file: PathBuf,
+}
+
+impl BenchSubprocess {
+    /// Waits for the process to finish and returns the measured instruction count
+    fn wait_and_get_instr_count(mut self) -> anyhow::Result<u64> {
+        let status = self
+            .process
+            .wait()
+            .context("Failed to run benchmark in cachegrind")?;
+        if !status.success() {
+            anyhow::bail!(
+                "Failed to run benchmark in cachegrind. Exit code: {:?}",
+                status.code()
+            );
+        }
+
+        let instruction_count = parse_cachegrind_output(&self.output_file)?;
+        std::fs::remove_file(&self.output_file).ok();
```

Comment:
would there be any value in an option to leave this output file in place for manual inspection?

Reply:
I think that would only be necessary for local debugging if things break in weird ways. That sounds improbable, so I'd rather have the developer comment out the line of code than adding an option to the CLI. Or do you envision another use case than debugging?

Reply:
Nope, that makes sense to me. Just spitballing :-)